include/deal.II/distributed/cell_weights.h

```diff
@@ -0,0 +1,170 @@
+     *
+     * An object of this class needs to exist for every DoFHandler associated
+     * with the Triangulation we work on to achieve satisfying work balancing.
+     *
+     * A small code snippet follows explaining how to achieve each cell
+     * being weighted by its current number of degrees of freedom.
+     * @code
+     * parallel::distributed::CellWeights<dim, spacedim>
+     *   cell_weights(hp_dof_handler);
+     * cell_weights.register_ndofs_weighting();
+     * @endcode
+     * The weighting function can be changed anytime. Even more ambitious
+     * approaches are possible by submitting customized functions, e.g.
+     * @code
+     * cell_weights.register_custom_weighting(
+     *   [](const FiniteElement<dim, spacedim> &) -> unsigned int {
+     *   return std::pow(fe.dofs_per_cell, 1.6);
+     * });
+     * @endcode
+     * @note Keep in mind that the return values are restricted to unsigned integers.
+     *
+     * @note Be aware that this class registers the weight function to the
+     * Triangulation's signal during its construction. If the Triangulation
+     * associated with the DoFHandler changes during the lifetime of the
+     * latter, an assertion will be triggered in the weight_callback function.
+     * It is recommended to create a separate object in this case and to destroy
+     * previous one.
+     *
+     * @ingroup distributed
+     * @author Marc Fehling, 2018
+     */
+    template <int dim, int spacedim = dim>
+    class CellWeights
+    {
+    public:
+      /**
+       * Constructor.
+       *
+       * @param[in] dof The hp::DoFHandler which will be used to
+       *    determine each cell's degrees of freedom.
+       */
+      CellWeights(const dealii::hp::DoFHandler<dim, spacedim> &dof_handler);
+
+      /**
+       * Destructor.
+       */
+      ~CellWeights();
+
+      /**
+       * Choose a constant weight on each cell.
+       */
+      void
+      register_constant_weighting();
+
+      /**
+       * Choose a weight reflecting the number of degrees of freedom on each
+       * cell.
+       */
+      void
+      register_ndofs_weighting();
+
+      /**
+       * Choose a weight reflecting the number of degrees of freedom
+       * <i>squared</i> on each cell.
+       */
+      void
+      register_ndofs_squared_weighting();
+
+      /**
+       * Register a custom cell weight by providing a function as a parameter.
+       */
+      void
+      register_custom_weighting(
+        std::function<unsigned int(const FiniteElement<dim, spacedim> &)>
+          custom_function);
+
+    private:
+      /**
+       * Pointer to the degree of freedom handler.
+       */
+      SmartPointer<const dealii::hp::DoFHandler<dim, spacedim>, CellWeights>
+        dof_handler;
+
+      /**
+       * Pointer to the Triangulation associated with the degree of freedom
+       * handler.
+       *
+       * We store both to make sure to always work on the correct combination of
+       * both.
```

Comment:
I don't understand. Can't you just get the triangulation from the dof handler whenever you need it?

Reply:
There is a member function called `initialize` that lets you assign a new `Triangulation` and `FECollection` to your `hp::DoFHandler` [(see doc)](https://www.dealii.org/9.0.0/doxygen/deal.II/classDoFHandler.html#aeca4a4af5c4d7f7c68a4def9ae87bc7b). That is the whole reason behind storing the triangulation separately and the assertion.

Reply:
Oh, I see.